### ISSUE TITLE
Proposal for improving rate, increase, delta (based on xrate, xincrease)

### ISSUE THREAD

**Issue by ColinDKelley** [prometheus/prometheus#12967]
### Proposal

Building on the excellent work of [Alin Sinpalean](https://github.com/free) discussed in https://github.com/prometheus/prometheus/issues/3806 and https://github.com/prometheus/prometheus/issues/12386, I would like to propose a variation that makes these improvements:

- results are linear (meaning that the sum of the increase over a time interval is the same across all possible query intervals);
- integer counters always return integer increases for all possible query intervals;
- the first counter increase from the origin is counted rather than ignored;
- missed or late scrapes never lead to missing results;
- there is no minimum query interval

Here is a Design Doc with a detailed explanation including examples: [Prometheus yrate (yrate, yincrease, ydelta): Linear versions of rate(), increase() and delta(), based on xrate()](https://docs.google.com/document/d/1CF5jhyxSD437c2aU2wHcvg88i8CjSPO3kMHsEaDRe2w/edit#heading=h.hzsa87ps5uhr)

We have been running a fork of Prometheus (and Thanos) for about 6 months now that has these improvements and it has been working great for us. That was important because we had run into all of the above limitations since converting our metrics stack from Graphite to Prometheus. I would be happy to submit a Pull Request with that code if there's interest.

**Comment by candlerb** [comment]
I like a lot about this, particularly the composability, but I can give some problematic cases - and hence yrate/yincrease is not necessarily a superior replacement for all previous options.

#### Counter origin

> For counters (yincrease, yrate), the implied [origin] value of every counter is 0. Although Prometheus never writes this 0 value, we can infer it when we find no samples in the lookbackInterval (default 5 minutes).

The essential problem here is that a target may go away for more than 5 minutes, whilst persisting its counter state internally. With the logic given here, you'd get a huge, spurious rate spike on the first sample after reconnecting to the target, as yrate thinks the value has rapidly increased from 0 to N.

A related problem is that the target may have counter state from before prometheus itself was started. The specific example I'm think of is network devices with SNMP counters like ifHCInOctets; if the device has been running for (say) a year already and then you point prometheus at it, the first counter value seen will be very high.

Clearly the counter must have been zero at some point in time previously, but there's no data point in the TSDB for it. (And if there were, it would be more than lookbackInterval ago - it could even be outside of the TSDB retention period).

I suppose each counter could have a related start time metric - but since counters don't have multiple parts like histograms, that would have to be a separate metric that you'd explicitly cross-reference[^1].

In short, I think that unless you write an explicit zero value for the origin when a counter is created, you can't assume the previous value was zero in recent history[^2].

However, removing this doesn't break the rest of the proposal; the claimed properties still apply, just not at the very start of the timeseries.

#### Rate accuracy across counter resets

Another issue is around counter resets which occur *within* a query range (your "restartSkew"). I'm thinking aloud here.

It's sounds reasonable to assume that they did reset to zero. That is, the sequence

```
   100 102   3   4   5
     |   |   |   |   |
     t0  t1  t2  t3  t4
```

is proposed to be treated as

```
   100 102 0 3   4   5
     |   |   |   |   |
     t0  t1  t2  t3  t4
```

However, it's arguable whether that gives more accurate results. You will catch some counts (the counter going from 0 to 3), but you will miss others (you don't know whether the counter went from 102 to 104, say, before it reset).

You say this is "unchanged" but I'm not sure it is. I haven't checked the source, but I thought the current rate logic is to take the known increases (102 - 100 = 2, 5 - 3 = 2, total known increase = 4) and divide them by the times where those increases happened, (t1-t0)+(t4-t2).  This gives an accurate rate for the time periods where the counter definitely increased from X to Y, whilst ignoring the time period between t1 and t2 when the increase in the counter value is unknown. We know it's at least 3, but it could be more.

For `increase()`, I can see why you'd want to include that count from 0 to 3.  But it still cannot know the count from 102 to 104 (or whatever), so what you get is a lower bound estimate.  The current logic of multiplying `rate()` by period gives a *possibly* closer estimate to the true value (if the rate is roughly constant), albeit one which is not integral.

One option is to assume (guess) that the counter reset occurred exactly half-way between the two scrapes, and that it was increasing at a fixed rate. That means if you see a value of 3, you set resetSkew to 6 (in the example above, you guess it went from 102 to 105 and then 0 to 3).  However this does meant that yincrease could return an over-estimate as well as an under-estimate of the number of counted events.  I can see there are many cases where you want to know for sure that *at least* the counted number of events had occurred.  And this option adds complexity for users who want to understand what the function does.

On balance, I think `yincrease()` counting only known increases is what people will intuitively expect, and they will accept that some may have been missed across counter resets. However if `yrate()` is defined to be `yincrease()` divided by query range, it could be (slightly) less accurate than what we have today.

--------
[^1] e.g. for SNMP devices, you can ask the device when it was powered up (subtract sysUpTime from current time). But note that any counter could have wrapped around one or more times during that time.

[^2] What timestamp would you assign to this zero value, and does it matter? For example:

```
(0)    [  2   8  15  22  ]
 |     |  |   |   |   |  |
(tN)   tX t0  t1  t2  t3 tY
```

The query range window, e.g. `foo[5m]`, is from tX to tY.  tN is the timestamp of the implied origin zero, which is not actually known in the absence of a counter start time metric.

When calculating the rate, the most accurate answer would be (22 - 0) / (t3 - tN).  Since tN isn't known, I think you're using (22 - 0) / (tY - tX). This is not unreasonable, but that implies that the counter came to life with zero value a whole scrape interval before the first actual data point. In reality, it could have come to life just before t0.  It's not a big deal; the first value of `yrate()` may be skewed, but the `yincrease()` will be unaffected.

**Comment by ColinDKelley** [comment]
Thanks for the quick and thoughtful feedback, @candlerb! I'll address each topic in a separate post here.

**Counter origin**

> The essential problem here is that a target may go away for more than 5 minutes, whilst persisting its counter state internally. With the logic given here, you'd get a huge, spurious rate spike on the first sample after reconnecting to the target, as yrate thinks the value has rapidly increased from 0 to N.

Although in our experience (about 6 months) we have not seen such a spike happen, you are absolutely correct that it could. However, we must contrast the risk of this spike problem against the certainty of the opposite problem, which we experienced every day while using the standard Prometheus `increase` function. This opposite problem I'm referring to is "missing counter origin increase". This arises each time a new time series is created. That might seem rare to some, but it's actually pretty common in our experience since Prometheus time series are uniquely identified by their labels, meaning that whenever a label changes, a new time series is created. I'll illustrate with an example that is very common in kubernetes, where we use the `pod` name as a label.

Consider these samples for the `webRequests` metric taken at a 15-second scrape interval, with a single label of the `pod`, which initially is `web-5q47s`:
```
timeSereies: "webRequests" { pod: "web-5q47s" }
samples: [ 3    8    15    21   21]
```
Then the pod restarts as `web-2z84a` and we continue to serve webRequests:
```
timeSereies: "webRequests" { pod: "web-2z84a" }
samples: [ 8    9    13    19   19]
```
(I've shown an unchanged sample values at the end so we don't get distracted by missing sample increases at the end.)

Now let's run the PromQL query with the minimum range it allows, which is double our sample interval:
```
increase(webRequests[30s])
```
(For this example we'll assume no jitter in the sample timestamps, so that `increase` won't do any extrapolation.) The values returned will be:
```
[ 12  6   5   6 ]
```
That is, it will see a total increase on the first pod of 12 + 6 = 18, and on the second pod of 5 + 6 = 11, vs the actual total increases that were 21 and 19.) The missed increases were 3 and 8.

**`increase` always misses the initial increase from the origin.**

> In short, I think that unless you write an explicit zero value for the origin when a counter is created,

I agree that writing an explicit 0 at the start of each counter when it's created in the TSDB would completely solve this with no corner cases. And I'm guessing that change would be not be disruptive to any existing code.

> However, removing this doesn't break the rest of the proposal; the claimed properties still apply, just not at the very start of the timeseries.

Agreed!

**Comment by ColinDKelley** [comment]
**Rate accuracy across counter resets**

> You say this is "unchanged" but I'm not sure it is.

I think so. I first learned of this behavior from Brian Brazil's presentations, such as [this one](https://www.slideshare.net/brianbrazil/counting-with-prometheus-cloudnativeconkubecon-europe-2017), on slide 25.

> If we see the value decreasing, then that's a counter reset so presume it went to 0.
> So seeing 0 -> 10 -> 5 is 10 + 5 = 15 of an increase.

I can confirm that PromQL `increase` and `rate` do exhibit this behavior.

And [here's the code](https://github.com/prometheus/prometheus/blob/main/promql/functions.go#L114-L121) that does that in the main branch. When the sample value steps backwards, a `0` intermediate value is inferred and that difference from 0 is added into the result on line 118:
```golang
resultFloat += prevValue
```

I agree with your comments about where this is imperfect. But I'm not proposing any changes to it here.

**Comment by candlerb** [comment]
Thank you for correcting me. I've also confirmed this with `promtool check rules`:

```
evaluation_interval: 1m

tests:
  - input_series:
      - series: foo
        values: '54 56 58 10 12 14'
    interval: 1m

    promql_expr_test:
      - expr: last_over_time(foo[5m])
        eval_time: 5m30s
        exp_samples:
          - value: 14
            labels: foo
      - expr: avg_over_time(foo[5m])
        eval_time: 5m30s
        exp_samples:
          - value: 30  # (56+58+10+12+14)/5
            labels: ""
      - expr: rate(foo[5m])
        eval_time: 5m30s
        exp_samples:
          - value: .06666666666666667  # increase 2+10+2+2 = 16, divided by 240
            labels: ""
```

(Aside: I was also interested to discover that both `avg_over_time` and `rate` drop the metric name from their result)

**Comment by ColinDKelley** [comment]
Wow, `promtool check rules` looks really helpful for this discussion. I didn't know about it but I'll read up on it!

**Comment by candlerb** [comment]
> We must contrast the risk of this spike problem against the certainty of the opposite problem, which we experienced every day while using the standard PromQL `increase` function. This opposite problem I'm referring to is "missing counter origin increase". This arises each time a new time series is created.

Indeed, your situation has the following characteristics:
* each timeseries starts when the pod starts
* the counter is zero when the pod starts
* prometheus is already running before the pod starts
* you don't want to miss the initial count

However for me, a timeseries starting when the counter is already very high (i.e. the SNMP counter example) is very common - either because the device was running before prometheus started, or because connectivity to the device was lost for more than lookbackInterval and then re-established. [^1]

I did wonder about having a user-defined threshold for the first value seen, for example:

`yrate(ifHCInOctets[5m], 1e10)`

This would accept a value of 1e10 or less with no preceding value as having a zero origin in front of it, but anything higher is treated as having no preceding origin.

The threshold value would be very application-specific. In this case, I'm thinking of a gigabit interface which could carry about 7.5e9 bytes in 1 minute, and anything much higher than that is likely to be spurious. For the pod counters you describe, you could give a very high upper bound, or a threshold of +Infinity would mean that *any* value seen could be the work done within one scrape of the pod startup.

However, including a specific threshold in every query would be very tedious.  If the threshold is omitted, it would have to be decided if the default should be +Infinity (your use case) or 0 (avoids any possibility of spurious spikes) - or even to have two functions, say `yrate` and `zrate`.

In the end, I think it's simpler to use the implicit zero origin for `yincrease`, but not for `yrate`, and then the default behaviour is the least surprising. You can always use the other function and multiply by a scale factor, if you want the other behaviour.

-------
[^1] Another example is many of the metrics from node_exporter, like `node_cpu_seconds_total` or `node_disk_io_time_weighted_seconds_total`. You need to take a `rate()` of these for meaningful results, but they have been accumulating since the target system started.

**Comment by ColinDKelley** [comment]
> In the end, I think it's simpler to use the implicit zero origin for yincrease, but not for yrate

Hmm, I wouldn't want for `yincrease` and `yrate` to stray from the relationship that `increase` and `rate` have in PromQL;  in fact I'd like to have the `y*` functions is to be drop-in replacements for the base PromQL functions. For those, `rate()` is simply `increase()/seconds`. So I made that [Design Rule 6](https://docs.google.com/document/d/1CF5jhyxSD437c2aU2wHcvg88i8CjSPO3kMHsEaDRe2w/edit#bookmark=id.fjbtgukwdohq) in the `yrate` Design Proposal.

_Source:_
- The [official rate docs](https://prometheus.io/docs/prometheus/latest/querying/functions/#rate) say that in the opening sentence and `rate` is [commonly explained that way](https://stackoverflow.com/questions/66674880/understanding-of-rate-function-of-promql).
- The [PromQL source](https://github.com/prometheus/prometheus/blob/main/promql/functions.go#L166-L168) is clearly written that way.

**Comment by free** [comment]
**Counter origin**

Strongly agree with @candlerb regarding implied counter origins. Targets appear and disappear together with their counters; some counters are persisted across target restarts; this behavior may work well for you, but should not be generalized across all counters.

At most, you could take inspiration from the existing `rate()` implementation (which uses heuristics to decide when to extrapolate and when not to) and only use an implied origin of zero if the initial value is within some epsilon of the average increase between the other samples (e.g. imply an initial value of zero for `[3, 8, 15, 21]`, but not for `[1003, 1008, 1015, 1021]`). But my personal choice would be to keep things simple and forget about any fractional (relative to the scrape interval) increase before the first sample.

**Integer increases for integer counters**

This is a nice property to have (and it was [one of the variants I explored](https://github.com/prometheus/prometheus/issues/3806#issuecomment-376944209) with `xrate()`, V1), but if the `increase()` or `rate()` range is not an exact multiple of the scrape interval, you get very jarring aliasing, with the output jumping back and forth between two values when given a smoothly increasing input (take a look at the `15s`, `25s` and `35s` rates [here](https://free.github.io/xrate/Grafana%20%2D%20xrate%20%2D%20no%20range%20adjust%2C%20smooth%20counter%2C%20series%20start.png)).

Furthermore, using a time range shorter than the scrape interval may result in an increase/rate that greatly overestimates the actual increase/rate. E.g. given a counter `foo` that increases by 1 per second; scraped once a minute; both `rate(foo[1s])` and `increase(foo[1s])` will produce outputs of 60. That counter did not increase by 60 in one second.

IMHO this is not necessarily a deal breaker, but the fact that it basically requires the use of multiples of the scrape interval should definitely be kept in mind,

**Outputting zero when no samples fall within the range**

To some extent this is implied by the above (because you want `avg_over_time(rate(foo[1s])[10m:1s])` to produce the same output as `rate(foo[10m])`). But while it may make sense to say that the increase in this situation is zero (because the values at the start and end of the range are both given by the same sample, so they're equal), one's mental image of the rate of increase (ignoring the fact that the definitions of PromQL's `rate()` and `increase()` are tightly coupled) is the change in value divided by the change in time, i.e. `0 / 0`. Which is not zero.

In other words, I believe it would be a lot less surprising to say that "because there's only one sample to go by, one cannot compute a rate of change" instead of "the rate of change is zero".

I for one (so what follows is my personal opinion) would be fine having an `increase()` function that outputs integers when fed integers (and either zero or no output when fed a single sample); and a `rate()` function that is not directly related to `increase()` that only outputs something when it has at least 2 samples to work with; and that output is the change in value divided by the change in time (because there is no good reason to do otherwise).

**Comment by bboreham** [comment]
I wanted to mention there is work under way to add a 'created' timestamp to each series, which makes some parts of the problem easier.

Most of the work links to #6541.

**Comment by candlerb** [comment]
Interesting, although it only helps if every exporter is modified to expose `foo_created` alongside every `foo` counter.

In the case of SNMP, this would mean synthesising many pseudo-metrics all derived from sysUpTime and the current time:

```
ifHCInOctets{ifAlias="gi0"} ...
ifHCInOctets_created{ifAlias="gi0"} ...
ifHCInOctets{ifAlias="gi1"} ...
ifHCInOctets_created{ifAlias="gi1"} ...
ifHCInOctets{ifAlias="gi2"} ...
ifHCInOctets_created{ifAlias="gi2"} ...
```

This also assumes that it's *not* required to update the _created metric every time a counter wraps (which would be impossible with SNMP)

**Comment by ColinDKelley** [comment]
For those who'd like to try it, we have a branch of 2.39.2 that supports `yincrease/yrate/ydelta` here:

https://github.com/Invoca/prometheus/tree/invoca-2.39.2-extensions

And here's a comparison pre-PR:
https://github.com/prometheus/prometheus/compare/v2.39.2...Invoca:prometheus:invoca-2.39.2-extensions

The core of the changes are here, in `func yIncrease` starting at promql/functions.go:220:

https://github.com/prometheus/prometheus/compare/v2.39.2...Invoca:prometheus:invoca-2.39.2-extensions#diff-d8538d074889981b3d2c55b012f620e45508e23b133996c7ca0b0594dd95e65f

If you do try this branch, you can set this environment variable to have the standard `increase/rate/delta` functions replaced with their `y*` counterparts. That's how we've been running it at our company for almost a year now.

```sh
REPLACE_RATE_FUNCS=2
```

Pro tip: you might see that `REPLACE_RATE_FUNCS=y` is supported in the code too...but that's a trap with some versions of YAML that convert `y` into a boolean `true`. That's been patched in `psych`, but I don't recall which versions had the bug so it's best just to avoid.

**Comment by 42Network** [comment]
Hello. I have a total newbie question for you. Is there an easy way I can apply this patch locally when I am just using Docker Compose to pull down the latest DockerHub Prometheus image?

If not, I would love to be pointed to a concise How To document on how to git clone your version and turn it into my own DockerHub image so other people on my team can still use Docker Compose to utilize this excellent patch.

Snippet from my docker-compose.yaml file:
```
#### prometheus ####
  prometheus:
    image: prom/prometheus:latest
    container_name: prometheus
    volumes:
     - ./prometheus:/etc/prometheus
     - prometheus_data:/prometheus
     - ./alertmanager/alert.rules:/alertmanager/alert.rules
    command: "--config.file=/etc/prometheus/prometheus.yml --storage.tsdb.path=/prometheus"
    ports:
     - 9090:9090
    restart: always
    depends_on:
#     - cadvisor
     - node-exporter
     - snmp-exporter
```
Thanks.

**Comment by beorn7** [comment]
Hello from the bug scrub!

The whole so-called "xrate issue" is well known and has been discussed countless times. The conclusion so far is that none of the alternatives have been proven to be a clear enough improvement to add them to Prometheus. Your design doc adds a few interesting aspects. If you like, you could transform it into a "proper" proposal via the [proposal process](https://github.com/prometheus/proposals), where more discussion could happen.

We'll close this issue here as we do not want single issues to track significant change proposals.
